src/main/java/org/opensearch/plugin/insights/core/utils/IndexedPriorityQueue.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.insights.core.utils;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class IndexedPriorityQueue<K, V> {
+
+    public static class Entry<K, V> {
+        public final K key;
+        public V value;
+
+        Entry(K key, V value) {
+            this.key = key;
+            this.value = value;
+        }
+    }
+
+    private final List<Entry<K, V>> heap;
+    private final Map<K, Integer> indexMap;
+    private final Comparator<? super V> comparator;
+
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    private final Lock read = lock.readLock();
+    private final Lock write = lock.writeLock();
+
+    public IndexedPriorityQueue(int initialCapacity, Comparator<? super V> comparator) {
+        this.comparator = comparator;
+        this.heap = new ArrayList<>(initialCapacity);
+        this.indexMap = new HashMap<>(initialCapacity);
+    }
+
+    public boolean insert(K key, V value) {
+        write.lock();
+        try {
+            if (indexMap.containsKey(key)) return false;
+            heap.add(new Entry<>(key, value));
+            int idx = heap.size() - 1;
+            indexMap.put(key, idx);
+            siftUp(idx);
+            return true;
+        } finally {
+            write.unlock();
+        }
+    }
+
+    public boolean remove(K key) {
+        write.lock();
+        try {
+            Integer idx = indexMap.remove(key);
+            if (idx == null) return false;
+
+            int lastIdx = heap.size() - 1;
+            if (idx != lastIdx) {
+                Entry<K, V> lastItem = heap.get(lastIdx);
+                heap.set(idx, lastItem);
+                indexMap.put(lastItem.key, idx);
+                siftDown(idx);
```

Comment:
Can we avoid calling both siftDown and siftUp in this case by doing some checks?